adcm_aio_client/core/requesters.py

```diff
@@ -22,20 +56,136 @@ def as_dict(self: Self) -> dict: ...
 
 
 class Requester(Protocol):
-    async def get(self: Self, path: str, query_params: dict) -> AsyncGenerator[RequesterResponse]: ...
+    async def login(self: Self, credentials: Credentials) -> Self: ...
+
+    async def get(self: Self, path: str, query_params: dict) -> RequesterResponse: ...
+
+    async def post(self: Self, path: str, data: dict) -> RequesterResponse: ...
+
+    async def patch(self: Self, path: str, data: dict) -> RequesterResponse: ...
+
+    async def delete(self: Self, path: str) -> RequesterResponse: ...
+
+
+@dataclass(frozen=True)
+class HTTPXRequesterResponse:
+    response: httpx.Response
+
+    def as_list(self: Self) -> list:
+        if not isinstance(self.json_data, list):
+            message = f"Expected a list, got {type(self.json_data)}"
+            raise RequesterResponseError(message)
+
+        return self.json_data
+
+    def as_dict(self: Self) -> dict:
+        if not isinstance(self.json_data, dict):
+            message = f"Expected a dict, got {type(self.json_data)}"
+            raise RequesterResponseError(message)
+
+        return self.json_data
+
+    @cached_property
+    def json_data(self: Self) -> Json:
+        try:
+            data = self.response.json()
+        except JSONDecodeError as e:
+            message = "Response can't be parsed to json"
+            raise RequesterResponseError(message) from e
+
+        if "results" in data:
+            data = data["results"]
 
-    async def post(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+        return data
 
-    async def patch(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
 
-    async def delete(self: Self, path: str) -> AsyncGenerator[RequesterResponse]: ...
+STATUS_ERRORS_MAP = {
+    400: BadRequestError,
+    401: UnauthorizedError,
+    404: NotFoundError,
+    409: ConflictError,
+    500: ServerError,
+}
 
 
-class Session: ...
+def convert_exceptions(func: Callable) -> Callable:
+    @wraps(func)
+    async def wrapper(*arg: Iterable, **kwargs: dict) -> httpx.Response:
```

Comment:
for unpackable arguments typehints are about exact item, not collection

these typehints means wrapper([], (), some={}, buddy={})

IIRC ParamSpec should be used for such cases (it will also help in typing input callables)

Reply:
refactored